File: src/main/java/store/wckd/server/service/JwtService.java
package store.wckd.server.service;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import store.wckd.server.entity.User;

import java.time.Instant;
import java.util.Date;

@Service
public class JwtService {
    @Value("${jwt.issuer}")
    private String issuer;

    private final UserService userService;
    private final Algorithm jwtAlgorithm;

    public JwtService(UserService userService, Algorithm jwtAlgorithm) {
        this.userService = userService;
        this.jwtAlgorithm = jwtAlgorithm;
    }

    public User decodeJwtToUser(String jwtString) {
        long id = 0;

        try {
            id = Long.parseLong(JWT.decode(jwtString).getSubject());
        } catch (Exception ignored) {
        }

        return userService.findById(id).block();
    }

    public String encodeJwt(User user) {
        return JWT.create()
                .withIssuedAt(Date.from(Instant.now()))
                .withIssuer(issuer)
                .withSubject(String.valueOf(user.getId()))
                .sign(jwtAlgorithm);
    }

}

File: src/main/java/store/wckd/server/service/JwtService.kt
package store.wckd.server.service

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import store.wckd.server.entity.User
import java.time.Instant
import java.util.*
import javax.persistence.EntityNotFoundException

@Service
class JwtService(
        private val userService: UserService,
        private val jwtAlgorithm: Algorithm
) {
    @Value("\${jwt.issuer}")
    private lateinit var issuer: String

    fun decodeJwtToUser(jwt: String): User =
            JWT.decode(jwt)
                    .subject
                    .toLongOrNull()
                    .let {
                        userService.findById(it ?: 0).block() ?: throw EntityNotFoundException()
                    }

    fun encodeJwt(user: User): String =
            JWT.create()
                    .withIssuedAt(Date.from(Instant.now()))
                    .withIssuer(issuer)
                    .withSubject(user.id.toString())
                    .sign(jwtAlgorithm)
}